be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp

```diff
@@ -0,0 +1,228 @@
+        state->runtime_filter_port()->publish_runtime_filters(bloom_filters);
+        // move runtime filters into RuntimeFilterHub.
+        runtime_filter_hub()->set_collector(_plan_node_id, std::make_unique<RuntimeFilterCollector>(
+                                                                   std::move(in_filters), std::move(bloom_filters)));
+    }
+    return Status::OK();
+}
+
+Status SpillableHashJoinBuildOperator::append_hash_columns(const ChunkPtr& chunk) {
+    auto factory = down_cast<SpillableHashJoinBuildOperatorFactory*>(_factory);
+    const auto& build_partition = factory->build_side_partition();
+
+    size_t num_rows = chunk->num_rows();
+    auto hash_column = spill::SpillHashColumn::create(num_rows);
+    auto& hash_values = hash_column->get_data();
+
+    // TODO: use different hash method
+    for (auto& expr_ctx : build_partition) {
+        ASSIGN_OR_RETURN(auto res, expr_ctx->evaluate(chunk.get()));
+        res->fnv_hash(hash_values.data(), 0, num_rows);
+    }
+    chunk->append_column(std::move(hash_column), -1);
+    return Status::OK();
+}
+
+bool SpillableHashJoinBuildOperator::is_finished() const {
+    return _is_finished;
+}
+
+Status SpillableHashJoinBuildOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
+    DeferOp update_revocable_bytes{
+            [this]() { set_revocable_mem_bytes(_join_builder->hash_join_builder()->hash_table().mem_usage()); }};
+
+    if (spill_strategy() == spill::SpillStrategy::NO_SPILL) {
+        return HashJoinBuildOperator::push_chunk(state, chunk);
+    }
+
+    if (!chunk || chunk->is_empty()) {
+        return Status::OK();
+    }
+
+    // TODO: materialize chunk (const/nullable)
+
+    RETURN_IF_ERROR(append_hash_columns(chunk));
+
+    RETURN_IF_ERROR(_join_builder->append_chunk_to_spill_buffer(state, chunk));
+
+    if (_is_first_time_spill) {
+        _is_first_time_spill = false;
+        _hash_table_slice_iterator = _convert_hash_map_to_chunk();
+        RETURN_IF_ERROR(_join_builder->append_spill_task(state, _hash_table_slice_iterator));
+    }
+
+    return Status::OK();
+}
+
+void SpillableHashJoinBuildOperator::mark_need_spill() {
+    if (!_is_finished) {
+        _join_builder->set_spill_strategy(spill::SpillStrategy::SPILL_ALL);
+    }
+}
+
+std::function<StatusOr<ChunkPtr>()> SpillableHashJoinBuildOperator::_convert_hash_map_to_chunk() {
+    auto build_chunk = _join_builder->hash_join_builder()->hash_table().get_build_chunk();
+    DCHECK_GT(build_chunk->num_rows(), 0);
+
+    _hash_table_build_chunk_slice.reset(build_chunk);
+    _hash_table_build_chunk_slice.skip(kHashJoinKeyColumnOffset);
+
+    return [this]() mutable -> StatusOr<ChunkPtr> {
+        if (_hash_table_build_chunk_slice.empty()) {
+            _join_builder->hash_join_builder()->reset(_join_builder->hash_table_param());
+            return Status::EndOfFile("eos");
+        }
+        auto chunk = _hash_table_build_chunk_slice.cutoff(runtime_state()->chunk_size());
```

Comment:
`_hash_table_build_chunk_slice` have more than one chunk?

Reply:
We don't want to spill the whole chunk to disk, I want to be able to split the task into smaller pieces